### PR TITLE
Choose join distribution type automatically

### PR DESCRIPTION
**Pull request by rschlussel-zz**
This is built on top of a branch that combines the statistics work with my work on determining the join distribution type.  This doesn't incorporate any of the recent suggestions for handling cost based optimizers, since that's not decided yet.  But this is what I've got according to the old model.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/metadata/GlobalProperties.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+public final class GlobalProperties
+{
+    private final DataSize maxMemoryPerNode;
+
+    public GlobalProperties()
+    {
+        this(new NodeMemoryConfig());
+    }
+    public GlobalProperties(NodeMemoryConfig nodeMemoryConfig)
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/metadata/GlobalProperties.java:29]: Why not just directly pass `NodeMemoryConfig` to optimizer which needs that?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DetermineJoinDistributionType.java [hunk 2]
```diff
@@ -132,18 +169,39 @@ private static boolean isCrossJoin(JoinNode node)
...
         {
-            return !getJoinDistributionType(session).equals(FeaturesConfig.JoinDistributionType.BROADCAST);
+            double smallSizeLimit = .1 * globalProperties.getMaxMemoryPerNode().toBytes();
+            Estimate dataSize = costCalculator.calculateCostForNode(session, node).getOutputSizeInBytes();
+            if (!dataSize.isValueUnknown()) {
+                return dataSize.getValue() < smallSizeLimit;
+            }
+
+            dataSize = costCalculator.calculateCostForNode(session, node).getOutputRowCount();
+            return !dataSize.isValueUnknown() && dataSize.getValue() * ROW_SIZE_ESTIMATE < smallSizeLimit;
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DetermineJoinDistributionType.java]: Leave it for now but IMO the cost calculator should do the estimation by itself if connector does not exose explicit stats. Then this section will not be necessary.

File: presto-cli/src/test/java/com/facebook/presto/cli/TestClientOptions.java [hunk 3]
```diff
@@ -143,27 +143,27 @@ public void testUpdateSessionParameters()
         ClientSession session = options.toClientSession();
         SqlParser sqlParser = new SqlParser();
 
-        ImmutableMap<String, String> existingProperties = ImmutableMap.of("query_max_memory", "10GB", "distributed_join", "true");
```
Comment (losipiuk) [inline presto-cli/src/test/java/com/facebook/presto/cli/TestClientOptions.java:146]: I guess we should keep the old property as deprecated at least for a while.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/JoinMatcher.java [hunk 4]
```diff
@@ -29,11 +29,18 @@
 {
     private final JoinNode.Type joinType;
```
Comment (losipiuk) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/JoinMatcher.java:31]: extract changes in matcher to separate commit.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/JoinMatcher.java [hunk 5]
```diff
@@ -29,11 +29,18 @@
...
     JoinMatcher(JoinNode.Type joinType, List<AliasPair> equiCriteria)
     {
+        this(joinType, equiCriteria, null);
+    }
+
+    JoinMatcher(JoinNode.Type joinType, List<AliasPair> equiCriteria, JoinNode.DistributionType distributionType)
+    {
         this.joinType = requireNonNull(joinType, "joinType is null");
         this.equiCriteria = requireNonNull(equiCriteria, "equiCriteria is null");
+        this.distributionType = distributionType;
```
Comment (losipiuk) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/JoinMatcher.java:44]: Optional instead nullable.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/SemiJoinMatcher.java [hunk 6]
```diff
@@ -27,19 +27,29 @@
     private final String sourceSymbolAlias;
     private final String filteringSymbolAlias;
     private final String outputAlias;
+    private final SemiJoinNode.DistributionType distributionType;
```
Comment (losipiuk) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/SemiJoinMatcher.java]: Optional instead nullable.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/JoinMatcher.java (hunk 5 above)
Comment (losipiuk) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/JoinMatcher.java:44]: Actually I think that for each field we want to test in matcher we should have a:
`Predicate<FieldType>` which we blindly test in `matches` method. We would set it to `Predicates.true()` for any field we do not care about.

cc: @kokosing

File: presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestDetermineJoinDistributionType.java [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+    private final CostCalculator costCalculator;
+
+    public TestDetermineJoinDistributionType()
+    {
+        this.queryRunner = new LocalQueryRunner(testSessionBuilder()
+                .setCatalog("local")
+                .setSchema("sf10")
+                .build());
+
+        automaticChoiceSession = Session.builder(queryRunner.getDefaultSession())
```
Comment (losipiuk) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestDetermineJoinDistributionType.java]: automaticJoinDistributionSession

File: presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestDetermineJoinDistributionType.java [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+                .setSystemProperty(SystemSessionProperties.JOIN_DISTRIBUTION_TYPE, "automatic")
+                .build();
+        partitionedJoinSession = Session.builder(queryRunner.getDefaultSession())
```
Comment (losipiuk) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestDetermineJoinDistributionType.java:58]: partitionedJoinDistributionSession

File: presto-cli/src/test/java/com/facebook/presto/cli/TestClientOptions.java (hunk 3 above)
Comment (rschlussel-zz) [reply presto-cli/src/test/java/com/facebook/presto/cli/TestClientOptions.java:146]: That's not really possible because they conflict with each other.   Unless the two properties are set to the same thing, it's really not clear what the correct behavior should be. Anything we choose would be confusing to the user.

Comment (losipiuk) [reply presto-cli/src/test/java/com/facebook/presto/cli/TestClientOptions.java:146]: It could be confusing. So for new users we should guide them to use new property. But it would be nice that old tools/scripts/... still work with new Presto release with new session property.

I think following semantics would work:
- we set default value of new property to "distributed" so it matches old property
- when determining distribution: 
  - if new property is set to some non-default value - use that
  - otherwise use value from old property.

Then hopefully in some time we will be able to remove old property after users migrate to new one.

File: presto-main/src/main/java/com/facebook/presto/metadata/GlobalProperties.java (hunk 1 above)
Comment (rschlussel-zz) [reply presto-main/src/main/java/com/facebook/presto/metadata/GlobalProperties.java:29]: I thought there might be other relevant properties (anyway that would need to be passed through the same way global properties is here... the point is that NodeMemoryConfig wasn't accessible to the optimizer).  Also, I wanted it to be clear that if something becomes a session property, that this is no longer the right way to call it.

File: presto-cli/src/test/java/com/facebook/presto/cli/TestClientOptions.java (hunk 3 above)
Comment (rschlussel-zz) [reply presto-cli/src/test/java/com/facebook/presto/cli/TestClientOptions.java:146]: Ok. That seems reasonable enough.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/JoinMatcher.java (hunk 5 above)
Comment (kokosing) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/JoinMatcher.java:44]: I am ok with either way suggested by Łukasz as long you don't use `null`.

**Comment by rschlussel-zz** [comment]
addressed comments.  changes have been squashed in so that I could redo the commits

**Comment by rschlussel-zz** [comment]
oh but i still have to combine this with lukasz's optimizer.  you can hold off on review for now.

**Comment by rschlussel-zz** [comment]
updated

File: presto-cli/src/test/java/com/facebook/presto/cli/TestClientOptions.java (hunk 3 above)
Comment (rschlussel-zz) [reply presto-cli/src/test/java/com/facebook/presto/cli/TestClientOptions.java:146]: Spoke to @KBP-TDC and Martin.  They said we should get rid of the old property altogether.  I'm going to change it back.

Comment (losipiuk) [reply presto-cli/src/test/java/com/facebook/presto/cli/TestClientOptions.java:146]: Cool. Thanks

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java [hunk 9]
```diff
@@ -91,15 +91,15 @@ public SystemSessionProperties(
...
-                        featuresConfig.isDistributedJoinsEnabled(),
-                        false),
-                booleanSessionProperty(
                         DISTRIBUTED_INDEX_JOIN,
                         "Distribute index joins on join keys instead of executing inline",
                         featuresConfig.isDistributedIndexJoinsEnabled(),
                         false),
+                stringSessionProperty(
+                        JOIN_DISTRIBUTION_TYPE,
+                        "What join method to use. Options are partitioned, replicated, or automatic",
```
Comment (akshatnair) [inline presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java]: What was your final decision to call this, partitioned or repartitioned? I saw in docs both the names being used. Lets pick one

File: presto-main/src/main/java/com/facebook/presto/execution/DataDefinitionExecution.java [hunk 10]
```diff
@@ -237,7 +238,8 @@ public String explain(Statement statement, List<Expression> parameters)
                 String query,
                 Session session,
                 Statement statement,
-                List<Expression> parameters)
+                List<Expression> parameters,
+                GlobalProperties globalProperties)
```
Comment (akshatnair) [inline presto-main/src/main/java/com/facebook/presto/execution/DataDefinitionExecution.java]: This argument seems to be unused? Do you need to pass it to DataDefinitionExecution?

File: presto-main/src/main/java/com/facebook/presto/metadata/GlobalProperties.java (hunk 1 above)
Comment (akshatnair) [reply presto-main/src/main/java/com/facebook/presto/metadata/GlobalProperties.java:29]: Are you defining global property as something that is defined when the server starts and a session property can override it?

**Review by akshatnair — COMMENTED** [review]
Do we have any tests for outer join and ON clause push down tests? It might be good to see if that works correctly after node flipping.

File: presto-main/src/main/java/com/facebook/presto/execution/DataDefinitionExecution.java (hunk 10 above)
Comment (rschlussel-zz) [reply presto-main/src/main/java/com/facebook/presto/execution/DataDefinitionExecution.java]: You're right.  It had been in the interface but it wasn't needed.

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java (hunk 9 above)
Comment (rschlussel-zz) [reply presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java]: renaming to repartitioned as discussed

File: presto-main/src/main/java/com/facebook/presto/metadata/GlobalProperties.java (hunk 1 above)
Comment (rschlussel-zz) [reply presto-main/src/main/java/com/facebook/presto/metadata/GlobalProperties.java:29]: Something that is defined when the server starts and there are no session properties to override it (like max-memory-per-node).  there was no access to any of these properties at planning time.

**Comment by rschlussel-zz** [comment]
Addressed comments

**Comment by akshatnair** [comment]
LGTM mod one testing question I asked about ON clause predicate pushdown!

**Comment by rschlussel-zz** [comment]
good point about the outer joins.  I think we don't handle it correctly since the predicate push down happens first.  Looking at it now.   

EDIT: We flip the left and right and the join type, so it's the same.  verified that it works correctly
